workload/workload.go

```diff
@@ -0,0 +1,200 @@
+	go func() {
+		http.Handle("/metrics", promhttp.HandlerFor(reg, promhttp.HandlerOpts{Registry: reg}))
+		log.Fatal(http.ListenAndServe(":2112", nil))
+	}()
+}
+
+// Setup uploads the documents generated by the workload, and calls the workloads Setup function
+func Setup(w Workload, numItemsArg int, scp *gocb.Scope, coll *gocb.Collection) {
+	numConc := 2000
+	workChan := make(chan DocType, numConc)
+	shutdownChan := make(chan struct{}, numConc)
+	var wg sync.WaitGroup
+
+	wg.Add(numConc)
+	for i := 0; i < numConc; i++ {
+		go func() {
+			for {
+				select {
+				case doc := <-workChan:
+					_, err := coll.Upsert(doc.Name, doc.Data, nil)
+					if err != nil {
+						panic(errors.Wrap(err, "Data load upsert failed."))
+					}
+				case <-shutdownChan:
+					wg.Done()
+					return
+				}
+			}
+		}()
+	}
+
+	// Create a random document using the given workload definition
+	for i := 0; i < numItemsArg; i++ {
+		workChan <- w.GenerateDocument(fmt.Sprintf("u%d", i))
+	}
+
+	// Call the worloads own Setup function to perform any workload specific setup
+	err := w.Setup()
+	if err != nil {
+		panic(errors.Wrap(err, "failed to setup workload"))
+	}
+}
+
+func Run(w Workload, numUsers int, runTime time.Duration) {
+	sigCh := make(chan os.Signal, 10)
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	go func() {
+		<-sigCh
+		cancelFn()
+	}()
+
+	// Signal handler for SIGTERM
+	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)
+
+	// Create a work group of goroutine runners sharing the same probabilities.
+	var wg sync.WaitGroup
+
+	wg.Add(numUsers)
+	for i := 0; i < numUsers; i++ {
+		go runLoop(ctx, w.Probabilities(), w.Functions(), w.Operations(), runTime, i, &wg)
+	}
+
+	wg.Wait()
+}
+
+func runLoop(
+	ctx context.Context,
+	probabilities [][]float64,
+	functions map[string]func(context.Context, Runctx) error,
+	operations []string,
+	runTime time.Duration,
+	runnerId int,
+	wg *sync.WaitGroup) {
+
+	// Current operation index
+	currOpIndex := 0
+
+	rng := rand.NewSource(int64(RandSeed + runnerId))
+	r := rand.New(rng)
+
+	timeout := time.After(runTime)
+
+	slog := zap.L().Sugar()
+
+	slog.Debugf("Starting runner %d…", runnerId)
+
+	var runCtx Runctx
+	runCtx.r = *r
+	runCtx.l = *zap.L() // TODO: create a log for results
+	// todo: move this into context.value, a KV store for junk
+
+	for {
+		select {
+		case <-ctx.Done():
+			slog.Debugf("Received cancel, stopping runner %d…", runnerId) // TODO: fix bug where only one runner stops
+			wg.Done()
+		case <-timeout:
+			slog.Debugf("Run time reached, stopping runner %d…", runnerId)
+			wg.Done()
+		default:
+			// Get the next operation index based on probabilities
+			nextOpIndex := getNextOperation(currOpIndex, probabilities, r)
+			// call the next function
+			nextFunction := operations[nextOpIndex]
+			slog.Debug(nextFunction)
+			attemptMetrics[nextFunction].Inc()
+
+			// sleep a random amount of time
+			t := r.Int31n(5000-400) + 400
```

Comment:
Reading this, I'm reminded why I put the sleep first.  I think my goal was to make sure we don't fire off a ton of work all at the same time and then have it gaussian after that.  

Might make sense to have some pause in the starting of these goroutines.

I'd also comment that I'd sort of like that we run for a bit before we start collecting metrics.  The early metrics are rarely valid and can have outliers that will confuse you later.  Like when you fork-bomb (goroutine bomb) yourself.  But, that's another PR…

Reply:
But, perhaps we can do this with prometheus after the fact?  Basically we can just analyze after a bit of time to look at only the steady state?

Reply:
I think your second suggestion would be the easiest thing to do, we could just shorten the window to not include the start of the workload.